app/services/course/material/zip_download_service.rb

```diff
@@ -2,11 +2,68 @@ class Course::Material::ZipDownloadService
   class << self
     # Downloads the materials and zip them.
     #
-    # @param [Course::Material::Folder] folder The folder containing the materials.
     # @param [Array<Course::Material>] materials The materials to be downloaded.
+    # @param [Course::Material::Folder] folder The folder containing the materials. This is the
+    #   lowest common ancestor folder of all the materials.
+    # @return [String] The path to the zip file.
+    def download_and_zip(materials, folder)
+      dir = download(materials, folder)
+      zip_dir(dir)
+    end
+
+    private
+
+    # Download materials and store them in a temporary directory.
+    #
+    # @param [Array<Course::Material>] materials The materials to be downloaded.
+    # @param [Course::Material::Folder] folder The folder containing the materials.
+    # @return [String] The path to the temporary directory.
+    def download(materials, folder)
+      materials = Array(materials)
+      tmp_dir = Dir.mktmpdir('coursemology-download-')
+      Dir.chdir(tmp_dir)
+      materials.each do |material|
+        download_material(material, folder)
+      end
+
+      tmp_dir
+    end
+
+    # Zip the directory and write to the file.
+    # @param [String] dir_path The path to the directory.
+    #
     # @return [String] The path to the zip file.
-    def download_and_zip(_folder, _materials)
-      # TODO: To implement
+    def zip_dir(dir_path)
+      output_file = dir_path + '.zip'
+      Zip::File.open(output_file, Zip::File::CREATE) do |zip_file|
+        Dir[File.join(dir_path, '**', '**')].each do |file|
+          zip_file.add(file.sub(File.join(dir_path + '/'), ''), file)
+        end
+      end
+
+      output_file
+    end
+
+    # Download the material and store it in the given parent directory.
+    def download_material(material, folder)
```

Comment:
As you said, if use an instance variable to store the base path, then instantiate the service object is better.